chat-android/src/main/java/com/ably/chat/Logger.kt

```diff
@@ -0,0 +1,108 @@
+package com.ably.chat
+
+import android.util.Log
+import java.time.LocalDateTime
+
+fun interface LogHandler {
+    fun log(message: String, level: LogLevel, throwable: Throwable?, context: LogContext)
+}
+
+data class LogContext(val tag: String? = null, val contextMap: Map<String, String> = mapOf())
+
+internal interface Logger {
+    val defaultContext: LogContext
+    fun withContext(additionalContext: LogContext): Logger
+    fun log(message: String, level: LogLevel, throwable: Throwable? = null, context: LogContext? = null)
+}
+
+internal fun Logger.trace(message: String, throwable: Throwable? = null, context: LogContext? = null) {
+    log(message, LogLevel.Trace, throwable, context)
+}
+
+internal fun Logger.debug(message: String, throwable: Throwable? = null, context: LogContext? = null) {
+    log(message, LogLevel.Debug, throwable, context)
+}
+
+internal fun Logger.info(message: String, throwable: Throwable? = null, context: LogContext? = null) {
+    log(message, LogLevel.Info, throwable, context)
+}
+
+internal fun Logger.warn(message: String, throwable: Throwable? = null, context: LogContext? = null) {
+    log(message, LogLevel.Warn, throwable, context)
+}
+
+internal fun Logger.error(message: String, throwable: Throwable? = null, context: LogContext? = null) {
+    log(message, LogLevel.Error, throwable, context)
+}
+
+internal fun LogContext.mergeWith(other: LogContext): LogContext {
+    return LogContext(
+        tag = other.tag ?: tag,
+        contextMap = contextMap + other.contextMap,
+    )
+}
+
+internal class AndroidLogger(
+    private val minimalVisibleLogLevel: LogLevel,
+    override val defaultContext: LogContext = LogContext(),
+) : Logger {
+
+    override fun withContext(additionalContext: LogContext): Logger {
+        return AndroidLogger(
+            minimalVisibleLogLevel = minimalVisibleLogLevel,
+            defaultContext = defaultContext.mergeWith(additionalContext),
+        )
+    }
+
+    override fun log(message: String, level: LogLevel, throwable: Throwable?, context: LogContext?) {
+        if (level.logLevelValue <= minimalVisibleLogLevel.logLevelValue) return
+        val finalContext = context?.let { defaultContext.mergeWith(it) } ?: this.defaultContext
+        val tag = finalContext.tag ?: "AblyChatSDK"
+
+        val contextString = if (this.defaultContext.contextMap.isEmpty()) "" else ", context: $finalContext"
+        val formattedMessage = "[${LocalDateTime.now()}] ${level.name} ably-chat: ${message}$contextString"
```

Comment:
Can we use formatting mention here -> https://github.com/ably-labs/ably-chat-kotlin/issues/20#issuecomment-2419626548.
Even in android, logs are sent to log analysers.

Reply:
https://betterstack.com/community/guides/logging/logging-microservices/